estimator/ntru_primal.py

```diff
@@ -0,0 +1,503 @@
+
+            NOTE: Non-overstretched parameters (where the probability of Dense sublattice
+            discovery is 0) will return β = d.
+
+            >>> params = NTRU.Parameters(n=113, q=512, Xs=ND.UniformMod(3), Xe=ND.UniformMod(3))
+            >>> NTRU.primal_dsd(params, red_shape_model="zgsa")
+            rop: ≈2^41.3, red: ≈2^41.3, δ: 1.012468, β: 42, d: 226, tag: dsd
+
+            >>> NTRU.primal_dsd(params, red_shape_model="cn11")
+            rop: ≈2^41.2, red: ≈2^41.2, δ: 1.012468, β: 41, d: 226, tag: dsd
+
+            >>> NTRU.primal_dsd(params, red_shape_model=Simulator.CN11)
+            rop: ≈2^41.2, red: ≈2^41.2, δ: 1.012468, β: 41, d: 226, tag: dsd
+
+        The success condition was formulated in [EC:KF17] and further refined in [AC:DvW21]
+        """
+        params = NTRUParameters.normalize(params)
+        # allow for a larger embedding lattice dimension: Bai and Galbraith
+        m = params.m + params.n if params.Xs <= params.Xe else params.m
+
+        try:
+            red_shape_model = simulator_normalize(red_shape_model)
+        except ValueError:
+            pass
+
+        if params.n > max_n_cache:
+            raise ValueError("Please increase the hardcoded value of max_n_cache to run the predictor for such large n")
+
+        remaining_proba = RR(1.)
+        average_beta = RR(0.)
+        total_DSD_prob = RR(0.)
+        DSD_prob = RR(0.)
+        prob_pos_total = np.zeros(2*params.n, dtype='double')
+
+        for beta in range(2, params.n):
+            tours = floor(params.n**2 / beta**2)+3
+
+            DSD_prob, DSD_prob_pos = self.prob_dsd(beta, params, red_shape_model, m=m,
+                                                   red_cost_model=red_cost_model, log_level=log_level)
+
+            if DSD_prob > 10e-8:
+                for t in range(tours):
+                    for i in range(2*params.n):
+                        prob_pos = DSD_prob_pos[i]
+                        average_beta += RR(beta) * remaining_proba * prob_pos
+                        prob_pos_total[i] += remaining_proba * prob_pos
+                        total_DSD_prob += remaining_proba * prob_pos
+                        remaining_proba *= (1.-prob_pos)
+
+                Logging.log("dsd", log_level+1, "β= %d,\t pr=%.4e, \t rem-pr=%.4e"%(beta, DSD_prob, remaining_proba))
+            if remaining_proba < 0.001:
+                average_beta += beta * remaining_proba
+                break
+
+        if not average_beta:
+            average_beta = m
+            predicate = False
+
+        else:
+            predicate = True
+
+        cost = costf(red_cost_model, RR(average_beta), m, predicate=predicate)
+        cost["tag"] = "dsd"
+        cost["problem"] = params
+        return cost.sanity_check()
+
+    __name__ = "primal_dsd"
+
+
+primal_dsd = PrimalDSD()
+
+
+class NTRUPrimalUSVP(PrimalUSVP):
+
+    @staticmethod
+    def _solve_for_d(params, m, beta, tau, xi):
```

Comment:
Is this copied from LWE? Can we reuse the code instead?

Reply:
Shall we make that a parameter?

Reply:
I think we could stick with the convention that tau=None means that the instance is homogeneous, but then we'd need to control how to set Tau to None. This could be done through an additional attribute in LWEParameters.